src/pygambit/game.pxi

```diff
@@ -994,48 +1014,79 @@ class Game:
-            raise UndefinedOperationError("append_move(): `actions` must be a nonempty list")
-        resolved_node.node.deref().AppendMove(resolved_player.player, len(actions))
-        for label, action in zip(actions, resolved_node.infoset.actions):
-            action.label = label
-
-    def append_infoset(self, node: typing.Union[Node, str],
+           raise UndefinedOperationError("append_move(): `actions` must be a nonempty list")
+        lock = threading.Lock()
+        if hasattr(nodes, '__iter__'):
+           with lock:
+              L = []
+              for nd in nodes:
+                 rs = cython.cast(Node, self._resolve_node(nd, 'append_move'))
+                 if rs in L:
+                    raise UndefinedOperationError("append_move(): `nodes` resolve to the same terminal node")
+                 if len(rs.children) > 0:
+                    raise UndefinedOperationError("append_move(): `node` must be a terminal node") 
+                 L.append(rs) 
+              for nd in nodes:
+                 resolved_node = cython.cast(Node, self._resolve_node(nd, 'append_move'))
+                 resolved_node.node.deref().AppendMove(resolved_player.player, len(actions))
+                 for label, action in zip(actions, resolved_node.infoset.actions):
+                    action.label = label
+        else:
+           resolved_node = cython.cast(Node, self._resolve_node(nodes, 'append_move'))
+           if len(resolved_node.children) > 0:
+              raise UndefinedOperationError("append_move(): `node` must be a terminal node")
+           resolved_node.node.deref().AppendMove(resolved_player.player, len(actions))
+           for label, action in zip(actions, resolved_node.infoset.actions):
+              action.label = label
+
+    def append_infoset(self, nodes: typing.Union[NodeReference, NodeReferenceSet],
                        infoset: typing.Union[Infoset, str]) -> None:
-        """Add a move in information set `infoset` at the terminal node `node`.
+        """Add a move in information set `infoset` at the group of terminal nodes `nodes`.
 
         Raises
         ------
         UndefinedOperationError
-            If `node` is not a terminal node.
+            If an element in `nodes` is not a terminal node.
         MismatchError
-            If `node` is a `Node` from a different game, or `infoset` is an `Infoset` from a
+            If an element in `nodes` is a `Node` from a different game, or `infoset` is an `Infoset` from a
             different game.
         """
-        resolved_node = cython.cast(Node, self._resolve_node(node, 'append_infoset'))
         resolved_infoset = cython.cast(Infoset, self._resolve_infoset(infoset, 'append_infoset'))
-        if len(resolved_node.children) > 0:
-            raise UndefinedOperationError("append_move(): `node` must be a terminal node")
-        resolved_node.node.deref().AppendMove(resolved_infoset.infoset)
-
+        if hasattr(nodes, '__iter__'):
```

Comment:
Same comments as above on duplication of code.  Further, nontrivial code is now being duplicated between `append_move` and `append_infoset` and one of the items in the spec was to ensure this didn't happen.